### PR TITLE
Playwright: Implement video capture.

### PR DESCRIPTION
**Pull request by worldomonation**
#### Changes proposed in this Pull Request

This patch implements video capture using Playwright 1.11.0.

- implements video capture using native Playwright API calls wrapped around hooks and additional processing. 
- uses a root-level `before` and `after` hook to start/terminate browser instances.

Implementation Details

- as of 1.11.0, Playwright is able to delete video files for a given Page object. It is also possible to provide a custom name when saving the video. However, my investigations point to there being a bug where `saveAs` writes two copies of the same file, one with original name and one with custom name. This requires additional processing in the case of failing tests, to manually remove the duplicate file.
- an array instantiated at the suite-level hook is used to track failures occurring in a given `mocha` run. Since we are using `magellan` in conjunction with `mocha`, each suite tagged with `@parallel` will execute in an isolated mocha environment. This means the array is guaranteed to contain only one item.

#### Testing instructions

Demo spec `wp-log-in-out-spec.js` is failing by default; ensure TeamCity is able to find and upload the recording to the Artifacts tab.

Partially implements #51082.
Parent: #51209
Child: #52843

### CODE REVIEW

**Comment by worldomonation** [comment]
The failures are from pinned chromedriver versions, it seems.

Will rebase on newer trunk and get rid of the failures.

File: packages/calypso-e2e/src/hooks/recording.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+ * Internal dependencies
+ */
+import { getTargetLocale, getTargetScreenSize } from '../browser-manager';
+import { getDateString, getVideoDir } from '../media-helper';
+
+export async function clearFailedTest( this: Context ): Promise< void > {
+	// String here would be more logical as combination of magellan + mocha should result
+	// in at most one failure per run of mocha.
+	// However, using a string here causes the recorded failures to not survive the context
+	// when the after hook is run whereas an array would.
```
Comment (scinos) [inline packages/calypso-e2e/src/hooks/recording.ts:16]: Out of curiosity, why an array works?

File: packages/calypso-e2e/src/hooks/recording.ts [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+/**
+ * Removes the file located by the path asynchronously then cleans the slate for next tests.
+ *
+ * @param {string} path Path-like string locating the object to be removed from disk.
+ * @returns {void} No return value.
+ */
+async function removeVideo( path: string ) {
+	unlink( path, function ( err ) {
```
Comment (scinos) [inline packages/calypso-e2e/src/hooks/recording.ts]: Nit: you may want to use [`unlink` from `fs/promises` ](https://nodejs.org/docs/latest-v15.x/api/fs.html#fs_fspromises_unlink_path) if you want to await for it to complete and keep consistent with other methods.

Same for `renameSync` a few lines below.

File: packages/calypso-e2e/src/hooks/recording.ts [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+export async function saveVideo( this: Context ): Promise< void > {
+	// Guard against page object not being present in this context.
+	if ( ! this.page ) {
+		return;
+	}
+
+	// Obtain the default save path that Playwright generates for the page.
+	const defaultVideoPath: string = await this.page.video().path();
+
+	// Close the BrowserContext so the video is written to disk.
```
Comment (scinos) [inline packages/calypso-e2e/src/hooks/recording.ts]: I'd suggest moving this to its own hook, and here thrown an error if the `BrowserContext` has not been closed yet.

**Review by scinos — APPROVED** [review]
Looks good! Added a few comments, but nothing blocking.

File: packages/calypso-e2e/src/hooks/recording.ts (hunk 1 above)
Comment (worldomonation) [reply packages/calypso-e2e/src/hooks/recording.ts:16]: I am not sure. I tried multiple times and could not pinpoint why a `string` does not and `array` preserves data.

File: packages/calypso-e2e/src/hooks/recording.ts (hunk 2 above)
Comment (worldomonation) [reply packages/calypso-e2e/src/hooks/recording.ts]: Thanks for pointing it out, the promise API is much more intuitive.

File: packages/calypso-e2e/src/hooks/recording.ts (hunk 3 above)
Comment (worldomonation) [reply packages/calypso-e2e/src/hooks/recording.ts]: I thought about this for a bit and I see why you suggest this - say browser context crashes, or is closed prematurely, then hook actions can fail or behave unpredictably. Is that the risk of simply leaving the code as is?

Comment (scinos) [reply packages/calypso-e2e/src/hooks/recording.ts]: My main concern was that nothing in `saveVideo` suggests the context will be closed. A developer may add another hook after `saveVideo` expecting the context to exist.

Comment (worldomonation) [reply packages/calypso-e2e/src/hooks/recording.ts]: Ah got it, thanks for clarifying! I will try to restructure this.

**Comment by worldomonation** [comment]
You know what's funny, Playwright released version 1.11.0 yesterday which has built-in support for the calls I've had to implement myself (rename, delete video).

**Pull request merged**